File: go/009-bridge/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

const (
	fileName = "input.txt"
)

type Point struct {
	X, Y    int
	History *Point
}

func (p *Point) areNeighbours(p2 *Point) bool {
	dX := math.Abs(float64(p.X - p2.X))
	dY := math.Abs(float64(p.Y - p2.Y))

	if dX > 1 {
		return false
	}
	if dY > 1 {
		return false
	}
	return true
}

func (p *Point) toString() string {
	x := strconv.Itoa(p.X)
	y := strconv.Itoa(p.Y)

	return fmt.Sprintf("[%s,%s]", x, y)
}

func (p *Point) move(dir string) *Point {

	switch dir {
	case "R":
		p = &Point{p.X + 1, p.Y, p}
	case "L":
		p = &Point{p.X - 1, p.Y, p}
	case "U":
		p = &Point{p.X, p.Y + 1, p}
	case "D":
		p = &Point{p.X, p.Y - 1, p}
	}
	return p
}

func moveThings(h *Point, tails []*Point, moves []string) map[string]bool {
	tailVisited := map[string]bool{
		"[0,0]": true,
	}
	for _, move := range moves {
		h = h.move(move)
		if !h.areNeighbours(tails[0]) {
			tails[0] = h.History
			// tails[0] = h.History
		}

		for i := 1; i < len(tails); i++ {
			if !tails[i].areNeighbours(tails[i-1]) {
				// tails[i].History = tails[i]
				tails[i] = tails[i-1].History
				// tails[i] = &Point{tails[i-1].History.X, tails[i-1].History.Y, tails[i]}
				// if i == 8 {
				// 	fmt.Println(tails[i].toString())
				// }
			}
		}
		tailVisited[tails[8].toString()] = true
		// fmt.Printf("%s %s %s %v\n", move, h.toString(), tails[0].toString(), tailVisited)
	}
	return tailVisited
}

func readFile(fileName string) []string {
	res := []string{}

	file, _ := os.Open(fileName)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, " ")

		num, _ := strconv.Atoi(parts[1])
		for i := 0; i < num; i++ {
			res = append(res, parts[0])
		}
	}
	return res
}

func main() {
	directions := readFile(fileName)

	fmt.Println(directions)
	tails := []*Point{}
	start := &Point{0, 0, nil}
	for i := 0; i < 10; i++ {
		tails = append(tails, start)
	}
	res := moveThings(tails[0], tails[1:], directions)
	fmt.Println(res)
	fmt.Println(len(res))
}

File: python/009-bridge/main.py
from collections import namedtuple

FILE_NAME="input.txt"

point = namedtuple("point", "x y")

def move_head(head: point, dir: str) -> point:
    if dir == "R":
        return point(head.x + 1, head.y)
    if dir == "L":
        return point(head.x - 1, head.y)
    if dir == "U":
        return point(head.x, head.y + 1)
    if dir == "D":
        return point(head.x , head.y - 1)

def is_neighbour(p1: point, p2: point) -> bool:
    return abs(p2.x - p1.x) <= 1 and abs(p2.y - p1.y) <= 1

def next_position(p1: point, p2: point) -> point:
    if is_neighbour(p1, p2):
        return p2
    
    if p1.x == p2.x:
        return point(p2.x ,(p2.y + p1.y)//2)
    
    if p1.y == p2.y:
        return point((p2.x + p1.x)//2 ,p2.y)
    
    diagonals = [[-1, -1], [1, 1], [-1, 1], [1, -1]]
    variants = list(map(lambda dia: point(p2.x + dia[0], p2.y + dia[1]), diagonals))
    for var in variants:
        if is_neighbour(p1, var):
            return var

def process_file(file_name):
    res = []
    with open(file_name, "r") as f:
        for line in f:
            dir, rep = line.split()
            for _ in range(int(rep)):
                res.append(dir)
    return res

def process_moves(head: point, tails: point, moves: list[str]): 
    visited1 = set()
    visited9 = set()

    for move in moves:
        head = move_head(head, move)
        tails[0] = next_position(head, tails[0])
        for i in range(1, len(tails)):
            tails[i] = next_position(tails[i -1], tails[i])
        visited1.add(tails[0])
        visited9.add(tails[8])
        print(head, tails[0])
    return len(visited1), len(visited9)


if __name__ == "__main__":
    input = process_file(FILE_NAME)
    print(input)
    tails = list(map(lambda _: point(0,0), range(9)))
    visited1, visited9 = process_moves(point(0,0), tails, input)
    print(visited1)
    print(visited9)